keps/sig-api-machinery/1904-efficient-watch-resumption/README.md

```diff
@@ -0,0 +1,502 @@
+  the resource version to which watch cache is synced from then on
+- if watch cannot be reestablished, all watchers connected to it are
+  disconnected, a new quorum list is send and watchcache is reset to
+  that point in time
+
+It's important to note that while only objects of a given resource type
+are watched (thus resource version to which watch cache is synced can
+only be updated to a value that some object of that type had), the quorum
+list, even though it is only requesting objects of that resource type,
+returns "globally current version for the whole etcd".
+
+While the inability to reestablish watch doesn't happen often in practice,
+the kube-apiserver restarts obviously does (e.g. on upgrades). And this
+is resulting in two main problems:
+
+1. In case of rolling upgrade of kube-apiserver, when no object of a given
+   resource type is changing (but objects of other types do), all watchers
+   will eventually be forced to relist, causing significant performance
+   and scalability issues for larger clusters. See the example below for
+   clarification.
+
+1. If no object of a given resource type is changing after kube-apiserver
+  startup, watch cache may stuck being synced to resource version not being
+	version of any object of a given type for extended period of time. See
+  https://github.com/kubernetes/kubernetes/issues/91073 for more details.
+
+To illustrate the problem better, imagine the following example:
+- there is a single kube-apiserver that we are going to restart
+- there are only two resource types: X and Y (for simplicity)
+- there is one object of type X: x with rv=100
+- there is one object of type Y: y with rv=101
+- there is a watch W for objects of type X already synced to rv=100
+- watchcache for both X and Y is up-to-date
+
+- T1: kube-apiserver being restarted
+- T2: kube-apiserver is initialized with watchcache for X synced to
+   rv=101 (most recent etcd version)
+- T3: watch W is trying to reconnect, given rv=100 is outside of
+   supported history window (we only support versions from 101 being
+   the moment of watch cache initialization), relist is forced
+
+Note that adding more kube-apiservers doesn't solve the problem. In
+fact, it is introducing the second problem where watch cache across
+kube-apiserver instances may be out of sync for extended period of time.
+
+To illustrate this problem imagine the following example:
+- the setup is as above, with the only difference of two kube-apiservers
+
+- T1: kube-apiserver-1 being restarted
+- T2: kube-apiserver-1 is initialized with watchcache for X synced to
+   rv=101 (most recent etcd version) [as above]
+- T3: object y is being updated to rv=102
+- T4: kube-apiserver-2 is being restarted
+- T5: kube-apiserver-2 is initialized with watchcache for X synced to
+   rv=102 (most recent etcd version)
+
+As long as no object of type X is being touched (created, updated or
+deleted), watchcache for X will not be updates. So even though they
```

Comment:
I still do not understand how this is different from the first problem. In both scenarios, for resource X, apiserver has an empty change window and believes X is synced to an RV number which a) is greater than any actual X object change's RV, and b) at which no X change happened.

In the second scenario there are multiple apiservers in the condition at the same time, at different RV numbers, but I don't see how that makes it worse.

I feel like there's some additional fact missing here, like if the RV given by list X from apiserver 2 can't be used to watch X on apiserver 1. But this doesn't seem like a big problem, since most list/watch calls will go to the same apiserver (and if you add a restart to break this, then it really is no different from problem 1). Additionally, for N apiservers in this condition at different RVs, the probability of a split-apiserver list/watch working is probably around 50% since the RV needn't exactly match? If it needed to exactly match the RV, the probability would be like N^-2 and that would get my attention :)

Reply:
> In the second scenario there are multiple apiservers in the condition at the same time, at different RV numbers, but I don't see how that makes it worse.

So the problem is that if you list from the newer kube-apiserver and then try to establish watch from that point in the older kube-apiserver you get "too old RV". And need to relist again. If you're unlucky, it may be indefinite...

Basically the problem is that you get different responses for the same request from two kube-apiservers for extended period of time, even though there isn't any network partition or anything like that.
I would say it's far from expected...